### PR TITLE
Add table lock to prevent race conditions

### PR DESCRIPTION
**Pull request by ajb**
This is probably not ideal for the core library, but I wanted to open this PR for continued discussion of https://github.com/djreimer/sequenced/issues/12. I first tried to wrap `#set_sequential_id` in an advisory lock (using https://github.com/mceachen/with_advisory_lock), but there was a less-frequent race condition where a different row would be created in between `#set_sequential_id` and `#save` being called.

### CODE REVIEW

**Comment by djreimer** [comment]
@ajb thanks for submitting this, I'll take a look

**Comment by ajb** [comment]
FWIW, we've been using this in production for a while now.

**Comment by hakunin** [comment]
This is how its done - locking records.
https://coderwall.com/p/h0qura/scope-sequences-in-activerecord
http://stackoverflow.com/questions/17024447/add-auto-increment-with-scope-to-existing-column-in-migration-file-rails

The first link is based on this gem.

I think it should be noted that this gem is not thread safe in the readme, since without thread safeti its pretty much useless for anyone.

File: lib/sequenced/generator.rb [hunk 1]
```diff
@@ -23,8 +23,10 @@ def skip?
     end
 
     def next_id
-      next_id_in_sequence.tap do |id|
-        id += 1 until unique?(id)
+      record.class.with_table_lock do
+        next_id_in_sequence.tap do |id|
+          id += 1 until unique?(id)
+        end
       end
```
Comment (samphilipd) [inline lib/sequenced/generator.rb]: This doesn't guarantee race-free behavior as far as I can tell.

Consider this scenario:

Thread 1 locks the table and uses this method to get the next id in the sequence.
Thread 1 releases the lock.
 - CONTEXT SWITCH -
Thread 2  locks the table and uses this method to get the next id in the sequence.
Thread 2 releases the lock.
Thread 2 inserts successfully.
- CONTEXT SWITCH - 
Thread 1 attempts to insert with the same value that thread 2 just used.

Result: disaster.

See my PR here for the correct place to put the lock https://github.com/djreimer/sequenced/pull/16 .

**Comment by ajb** [comment]
Thanks for the review! Closing in favor of #16, which looks :star2:

**Pull request closed without merging**